File: src/322. Coin Change.cpp
Problem link: https://leetcode.com/problems/coin-change/
Problem difficulty: Medium

/*** Solution Detail **
 * Time Complexity: O(amount * total number of coins)
 * Space Complexity O(amount) 
 * Approach: Dynamic Programming - Bottom up(Iterative)
 */ 
 
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        const int INF = 1e9;
        vector<int> dp(amount + 1, INF);
        dp[0] = 0;
        for (int x = 1; x <= amount; ++x) {
            for (auto &c : coins) {
                if (x - c >= 0 && dp[x - c] + 1 < dp[x])
                    dp[x] = dp[x - c] + 1;
            }
        }
        return (dp[amount] == INF ? -1 : dp[amount]);
    }
};

File: Leetcode/322. Coin Change.java
/*
Problem Link: https://leetcode.com/problems/coin-change/
Problem Type: DP, Recursion, caching  
*/

class Solution {
    private int calculate(int[] coins, int amount, int[] cache) {
        if (amount < 0) return -1;
        if (cache[amount] > 0 || amount == 0) return cache[amount];
        cache[amount] = 10001;
        for (int c : coins) {
            if (c > amount) continue;
            int remaining = calculate(coins, amount - c, cache);
            if (remaining < 0) {
                continue;
            }
            else {
                cache[amount] = Math.min(1 + remaining, cache[amount]);
            }            
        }
        return (cache[amount] < 10001 ? cache[amount] : -1);
    }

    public int coinChange(int[] coins, int amount) {
        int[] cache = new int[amount + 1];
        return calculate(coins, amount, cache);
    }
}
